File: src/main/java/intention_action/Snippets.java
package intention_action;

import helper.DataService;

public class Snippets {
    public static final DataService data = DataService.getInstance();

    public static final String PREFIX_SELECTION = "Subject";
    public static final String SUFFIX1 = data.logicName;

    public static final String GetX_SNIPPET_KEY = PREFIX_SELECTION + SUFFIX1;
    public static final String STATE_SNIPPET_KEY = PREFIX_SELECTION + SUFFIX1;


    static String getSnippet(SnippetType snippetType, String widget) {
        switch (snippetType) {
            case Obx:
                return snippetObx(widget);
            case GetBuilder:
                return snippetGetBuilder(widget);
            case GetX:
                return snippetGetX(widget);
            default:
                return "";
        }
    }

    private static String snippetObx(String widget) {
        return String.format("Obx(() {\n" +
                "  return %1$s;\n" +
                "})", widget);
    }

    private static String snippetGetBuilder(String widget) {
        return String.format("GetBuilder<%1$s>(\n" +
                "  builder: (%2$s) {\n" +
                "    return %3$s;\n" +
                "  },\n" +
                ")", GetX_SNIPPET_KEY, data.logicName.toLowerCase(), widget);
    }

    private static String snippetGetX(String widget) {
        return String.format("GetX<%1$s>(\n" +
                "  init: %1$s(),\n" +
                "  initState: (_) {},\n" +
                "  builder: (%2$s) {\n" +
                "    return %3$s;\n" +
                "  },\n" +
                ")", GetX_SNIPPET_KEY, data.logicName.toLowerCase(), widget);
    }
}

File: src/main/kotlin/intention_action/Snippets.kt
package intention_action

import helper.DataService
import java.util.*

object Snippets {
    const val PREFIX_SELECTION = "Subject"

    fun getSnippetKey(): String {
        val data = DataService.instance
        return PREFIX_SELECTION + data.module.logicName
    }

    fun getSnippet(snippetType: SnippetType?, widget: String): String {
        return when (snippetType) {
            SnippetType.Obx -> snippetObx(widget)
            SnippetType.GetBuilder -> snippetGetBuilder(widget)
            SnippetType.Observer -> snippetObserver(widget)
            SnippetType.GetBuilderAutoDispose -> snippetGetBuilderAutoDispose(widget)
            SnippetType.GetX -> snippetGetX(widget)
            else -> ""
        }
    }

    private fun logicName(): String {
        return DataService.instance.module.logicName
    }

    private fun snippetObx(widget: String): String {
        return String.format(
            """Obx(() {
  return %1${"$"}s;
})""", widget
        )
    }

    private fun snippetGetBuilder(widget: String): String {
        val key = getSnippetKey()
        return String.format(
            """GetBuilder<%1${"$"}s>(builder: (%2${"$"}s) {
  return %3${"$"}s;
})""", key, logicName().lowercase(Locale.getDefault()), widget
        )
    }

    private fun snippetObserver(widget: String): String {
        return String.format(
            """Observer(builder: (BuildContext context) {
  return %1${"$"}s;
})""", widget
        )
    }

    private fun snippetGetBuilderAutoDispose(widget: String): String {
        val key = getSnippetKey()
        return String.format(
            """GetBuilder<%1${"$"}s>(
  assignId: true,
  builder: (%2${"$"}s) {
    return %3${"$"}s;
  },
)""", key, logicName().lowercase(Locale.getDefault()), widget
        )
    }

    private fun snippetGetX(widget: String): String {
        val key = getSnippetKey()
        return String.format(
            """GetX<%1${"$"}s>(
  init: %1${"$"}s(),
  initState: (_) {},
  builder: (%2${"$"}s) {
    return %3${"$"}s;
  },
)""", key, logicName().lowercase(Locale.getDefault()), widget
        )
    }
}